clustered/server/src/main/java/org/ehcache/clustered/server/store/ServerStore.java

```diff
@@ -0,0 +1,113 @@
+ *
+ *  hash -> |seqNum1|payLoad| - |seqNum2|payLoad| - |seqNum3|payLoad|
+ *  hash -> |seqNum1|payLoad| - |seqNum2|payLoad|
+ *
+ *  This only allows append & replace operations rather than
+ *  normal map operations.
+ *
+ */
+public interface ServerStore {
+
+  /**
+   * Returns the Chain associated with the provided hash.
+   * An empty Chain if no mapping exists.
+   *
+   * @param key hashcode of the key
+   * @return the {@link Chain} associated with the hash
+   */
+  Chain get(long key);
+
+  /**
+   * Appends the provided binary to Chain associated with key atomically.
+   * While appending, the payLoad is stored in {@link Link} associated with
+   * a sequence number.
+   *
+   * @param key to which the payLoad has to be appended
+   * @param payLoad to be appended
+   */
+  void append(long key, byte[] payLoad);
+
+  /**
+   * The Chain associated with key, previous to append is returned.
+   * The operation atomically appends payLoad and returns the previously associated Chain
+   * with the key.
+   *  Following block of instructions are atomically performed.
+   *  {
+   *    Chain prevChain = get(key);
+   *    append(key, payLoad);
+   *    return prevChain;
+   *  }
+   *
+   * @param key to which the payLoad has to be appended
+   * @param payLoad to be appended
+   * @return the Chain associated with the key before payLoad was appended
+   */
+  Chain getAndAppend(long key, byte[] payLoad);
+
+  /**
+   * Replaces the provided Chain with the equivalent Chain present as per
+   * provided sequence Numbers for the provided key.
+   * This operation is not guaranteed to succeed.
+   * The replace is successful iff the Chain associated with the key has
+   * a sub-sequence of links present with the provided list of sequence numbers.
+   *
+   * If below mapping is present:
+   *
+   *    hash -> |1|payLoad| - |2|payLoad| - |3|payLoad|
+   *
+   * And replace(hash, |3|payLoad|, 1,2,3) is invoked then this operation will succeed & the
+   * final mapping would be:
+   *
+   *    hash -> |3|payLoad|
+   *
+   * Failure case:
+   *
+   * If below mapping was returned during get operation :
+   *
+   *    hash -> |1|payLoad| - |2|payLoad| - |3|payLoad|
+   *
+   * But before replace it was modified to :
+   *
+   *    hash -> |3|payLoad| - |4|payLoad|
+   *
+   * then replace(hash, |3|payLoad|, 1,2,3) will be ignored.
+   *
+   * Chain to be replaced, should also adhere to the rule of ascending order of Links,
+   * & all the sequence numbers in that chain should fall in the range of provided sequenceNumbersExpected
+   * as the sequence number is generated by server else operation will be ignored.
+   *
+   * @param key for which Chain has to be replaced
+   * @param update the Chain to be replaced for the provided sub-sequence
+   * @param sequenceNumbersExpected of the original Chain which this operation intends to replace
+   */
+  void replace(long key, Chain update, long... sequenceNumbersExpected);
```

Comment:
At this level we just want to expose the replace arguments as (long, Chain, Chain)... the fact that we're using sequence numbers internally should be hidden from the API consumer.